python/sensr_message_listener.py

```diff
@@ -39,78 +45,114 @@ def __init__(self,
         self._listener_type = listener_type
         self._output_ws = None
         self._point_ws = None
+        self._state = MessageListener.State.STOPPED
         
 
     def is_output_message_listening(self):
         return self._listener_type == ListenerType.OUTPUT_MESSAGE or self._listener_type == ListenerType.BOTH
 
     def is_point_result_listening(self):
         return self._listener_type == ListenerType.POINT_RESULT or self._listener_type == ListenerType.BOTH
+    
+    def check_oveflow_error(self, output):
+        if output.HasField('event') and output.event.HasField('health'):
+            if output.event.health.master == SystemHealth.Status.OUTPUT_BUFFER_OVERFLOW:
+                return True
+        return False
 
     async def _output_stream(self):
-        async with websockets.connect(self._output_address, ssl=self._ssl_context, compression=None, max_size=None, ping_interval=None) as websocket:
-            self._output_ws = websocket
-            while self._is_running:
-                try:
-                    message = await asyncio.wait_for(websocket.recv(), timeout=1.0) # Receive output messages from SENSR                    
-                    output = OutputMessage()
-                    output.ParseFromString(message)
-                    self._on_get_output_message(output)
-                except asyncio.TimeoutError:
-                    pass
-            # Clean up websocket connection
-            await self._output_ws.close()
-            # if point websocket is already closed, then call loop.stop()
-            # loop.stop should be called to exit `loop.run_forever`
-            if ((self._point_ws == None) or 
-                (self._point_ws != None and self._point_ws.closed == True)):
-                if (self._loop != None and self._loop.is_running()):
-                    self._loop.stop()
+       while self._state != MessageListener.State.STOPPED and self._state != MessageListener.State.STOP_REQUESTED:
+            async with websockets.connect(self._output_address, ssl=self._ssl_context, compression=None, max_size=None, ping_interval=None) as websocket:
+                if self._output_ws != websocket:
+                    self._output_ws = websocket
+                while not self._output_ws.closed:
+                    try:
+                        message = await asyncio.wait_for(self._output_ws.recv(), timeout=1.0) # Receive output messages from SENSR                    
+                        output = OutputMessage()
+                        output.ParseFromString(message)
+                        if self.check_oveflow_error(output):
+                            self._on_error("Output Buffer Overflow.")
+                        self._on_get_output_message(output)
```

Comment:
Do we need to continue execution after entering the if-statement? I.e. if the if-statement is entered do we also need to run `self._on_get_output_message(output)`?

Reply:
Yes, there can be a case which the actual client code (e.g. console_output.py) does not do anything in `_on_error` function.
In this case, sensr_message_listener still need to publish output message.